In [5]:
import os
import urllib.request
import re

In [2]:
if not os.path.exists("the-verdict.txt"):
    url = ("https://raw.githubusercontent.com/rasbt/LLMs-from-scratch/refs/heads/main/ch02/01_main-chapter-code/the-verdict.txt")
    file_path = "the-verdict.txt"
    urllib.request.urlretrieve(url, file_path)


In [ ]:
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
len(raw_text) 

20479

In [6]:
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', raw_text)
preprocessed = [item.strip() for item in preprocessed if item.strip()]
print(preprocessed[:30])

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in']


### Converting tokens to IDs

In [7]:
all_words = sorted(set(preprocessed))
vocab_size = len(all_words)

print(vocab_size)

1130


So we are left with the 1130 words/tokens instead of 20479 words/tokens, now we can do the mapping of these to the IDs for now

Below is a very simple method for tokenization

In [21]:
vocab = {val : idx for idx, val in enumerate(all_words)}
class SimpleTokenizer:
    def __init__(self, vocab):
        self.word2idx = vocab
        self.idx2word = {idx : word for word, idx in vocab.items()}
    
    def encode(self, text):
        preprocessed_text = re.split(r'([,.:;?_!"()\']|--|\s)', text)
        preprocessed_text = [item.strip() for item in preprocessed_text if item.strip()]
        idxs = [self.word2idx[word] for word in preprocessed_text]
        return idxs
    
    def decode(self, idxs):
        words = [self.idx2word[idx] for idx in idxs]
        return words

In [17]:
simp = SimpleTokenizer(vocab)

In [20]:
nums = simp.encode("Jack thought always genius")
# words = simp.decode(nums)

KeyError: 'J'